### DISCUSSION TITLE
How to mix fluids in the TH Module Single-Phase-Flow tutorial?

### DISCUSSION THREAD

**Discussion by yifanm4** [idaholab/moose#23132] [Q&A Modules: General] [answered]
Greetings All, 

I would like to use the 1-D single-phase-flow structure as the basic structure to implement graphite dust motion with the helium gas.

I understand that it has been made clearly that the single-phase-flow doesn't support mixing fluids, so my questions are:
1. Is there a way to manipulate the code to make it support the mix of the fluids? (The dust shall flow with the helium gas, thus it will still be a single phase flow)
2. Is this a situation where the Multi-App functionalities jump in? 
3. Is there a better module to solve this issue without taking too much computational power?

Thank you so much!

**Comment by GiudGiud** [comment, marked as answer]
Hello

1) yes. This will become a scalar transport problem. We do not support this right now in THM (@joshuahansel for fact-checking me) but I m very confident it's not difficult to add. 
2) You may elect to use MultiApps. The main reason for that is to limit the number of variables in a single solve, especially when the transported species does not influence the velocity distribution of the main fluid. 
3) we have similar problematics with the Navier Stokes module (transporting neutron precursors in molten salt reactor) and  in the future in the scalar transport module (this module will serve, in the future, as a base for tritium transport in the open-source TMAP8). If your system is small you can try finite volume Navier Stokes. 

Guillaume

**Reply by joshuahansel** [reply]
Fact-check returns TRUE (No scalar transport currently in THM).

**Comment by joshuahansel** [comment]
Note that we do have the following fluid properties formulation for arbitrary mixtures of gases: https://mooseframework.inl.gov/source/userobjects/IdealRealGasMixtureFluidProperties.html. There are alternative formulations out there.

After this piece, you can decide on what PDEs you want to be solving. It's possible that maybe you'll just be able to add an equation or two alongside the ones in `FlowChannel1Phase`.

**Reply by yifanm4** [reply]
Thank you so much for all these useful information! I am still new to Moose, how exactly do I add equations in FlowChannel1Phase?

**Reply by joshuahansel** [reply]
I'd just clarify that this wouldn't be "in FlowChannel1Phase", but would just be objects that you add in your input file alongside it.

As far as how to add equations, that's a pretty broad topic - you'll probably need to go through at least some tutorials: https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html. But the residual objects that will be relevant will be
- Kernels: https://mooseframework.inl.gov/syntax/Kernels/index.html
- DGKernels: https://mooseframework.inl.gov/syntax/DGKernels/index.html (depending on what spatial discretization you choose)
- BCs: https://mooseframework.inl.gov/syntax/BCs/index.html

But first, let us know which equations you want to solve so that we can check if there is something existing that you can use.

**Reply by yifanm4** [reply]
Hello Joshua, 

I might have gotten it wrong to begin with. Instead of mixing two types of fluids together, I actually want to use the calculated velocity (u) to couple with my source terms (graphite dust, fission products, and so on). By starting with the customized 1-D diffusion equation:

partial_f(x,t) /partial_t = u* partial_f(x,t) /partial_x + source, where f(x,t) could be graphite dust or fission products and u is the calculated velocity from single-phase-flow. 

Is there something existing that can help me with this? 

Yifan

**Reply by joshuahansel** [reply]
Ok, that seems like it shouldn't be too bad then. There's nothing in the TH module to solve this equation (and I don't think the scalar transport module would help you here yet), but I think it would be easy enough to implement - just need a kernel for a product of a variable and a gradient of another variable. There might be such a kernel already.

**Reply by yifanm4** [reply]
I tried to use the existing kernels (ConservativeAdvection) which seems to get the job done. However, since the velocity is not from the input parameters, but a calculated output instead. It gives me the error message that "coupled variable 'vel' was not found", what shall I do in this case?

**Reply by GiudGiud** [reply]
you can define an auxvariable vel and initialize it to the value you want.

**Reply by joshuahansel** [reply]
While `FlowChannel1Phase` should give you a variable called `vel`, I see that `ConservativeAdvection` is requesting a _vector_ variable, whereas `vel` is scalar (since it's 1-D). So you would need an aux kernel that takes `vel` and multiplies it by a vector (your normalized `orientation` vector of `FlowChannel1Phase`. The caution here is that you will lose the automatic differentiation when involving aux kernels, so your convergence may suffer. I'd recommend a kernel that uses a scalar `vel` directly. If one does not exist, you could write one (for now, you could copy and paste `ConservativeAdvection` and then make the changes there):
```
_velocity[_qp]
```
goes to
```
_vel[_qp] * _orientation
```

**Reply by yifanm4** [reply]
I located the **ConservativeAdvection**. But once I clicked in, there are just some input files. How can I change the source file for **ConservativeAdvection** as you described above?

**Reply by joshuahansel** [reply]
[ConservativeAdvectionScalar_h.txt](https://github.com/idaholab/moose/files/10459812/ConservativeAdvectionScalar_h.txt)
[ConservativeAdvectionScalar_C.txt](https://github.com/idaholab/moose/files/10459833/ConservativeAdvectionScalar_C.txt)

Rename these two files to `ConservativeAdvectionScalar.h` and `ConservativeAdvectionScalar.C`, respectively, and then put them into `framework/include/kernels` and `framework/src/kernels`, respectively. Recompile your executable and try it out.

**Reply by yifanm4** [reply]
I am trying to add my own AuxKernel object called DustConvectionAux to the TH tutorial. 

I create **DustConvectionAux.h** and **DustConvectionAux.C**, put them in ./modules/thermal_hydralics/include/auxkernel and ./modules/thermal_hydralics/src/auxkernel respectively. I register it to the ThermalHydraulicsApp as well, _but it still complains that **DustConvectionAux** is not registered._ 

I also tried to register it in my src/base/ThermalHydraulicsAppApp.C file, still not working. 

It can be complied just fine. 

Any idea about how to fix this?

**Reply by joshuahansel** [reply]
What executable did you compile/run?

**Reply by joshuahansel** [reply]
> I also tried to register it in my src/base/ThermalHydraulicsAppApp.C file, still not working.

You do not need to do anything to that file, just the register call in `DustConvectionAux.C`.

**Reply by yifanm4** [reply]
> What executable did you compile/run?

Just used thermal_hydraulics-opt* to run 01_flow_channel.i 

The DustConvection is the same as [Example 10: Auxiliary Variables](https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex10_aux.html), I just renamed the ExampleAux to DustConvection everywhere. 

I add the following AuxVariable and AuxKernel to 01_flow_channel.i 

[AuxVariables]
  [dust_conv]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxKernels]
  [nodal_example]
    type = DustConvection
    variable = dust_conv
    value = 3.0
    coupled = _vel
  []
[]

**Reply by yifanm4** [reply]
> > I also tried to register it in my src/base/ThermalHydraulicsAppApp.C file, still not working.
> 
> You do not need to do anything to that file, just the register call in `DustConvectionAux.C`.

I did try to register it to the Thermal Hydralic App: 

registerMooseObject("ThermalHydraulicsApp", DustConvectionAux);

**Reply by joshuahansel** [reply]
In your example, you have `type = DustConvection`, but your register call has `DustConvectionAux`. I just want to check that this was a typo in your comment and not in your files.

**Reply by yifanm4** [reply]
> In your example, you have `type = DustConvection`, but your register call has `DustConvectionAux`. I just want to check that this was a typo in your comment and not in your files.

Thanks for pointing it out. I did make a typo there, but it still makes no difference after I changed it to DustConvectionAux:

***** ERROR ***
A 'DustConvectionAux' is not a registered object.**

**Reply by joshuahansel** [reply]
This should not happen, based on what you've told me. If you are sure that you have recompiled `thermal_hydraulics-opt` and not some other executable, then the only thing I can offer at this point is that I look at your fork myself.

**Reply by joshuahansel** [reply]
And you did keep this line, right?
```
registerMooseObject("ThermalHydraulicsApp", DustConvectionAux);
```
I'm just making sure because you said:

> I just renamed the ExampleAux to DustConvection everywhere.

**Reply by yifanm4** [reply]
> And you did keep this line, right?
> 
> ```
> registerMooseObject("ThermalHydraulicsApp", DustConvectionAux);
> ```
> 
> I'm just making sure because you said:
> 
> > I just renamed the ExampleAux to DustConvection everywhere.

Yes, I kept this line.

**Reply by yifanm4** [reply]
> This should not happen, based on what you've told me. If you are sure that you have recompiled `thermal_hydraulics-opt` and not some other executable, then the only thing I can offer at this point is that I look at your fork myself.

You are correct. I think the issue occurs becomes I originally didn't want to mess up the original file, thus I copy the thermal_hydralics module and renamed it as thermal_hydralics_copy. All the changes I mentioned above you made in the thermal_hydralics_copy, and I suppose it's the reason why it doesn't work. Once I go into the actual thermal_hydralics folder and make the same changes, it compiled successfully.

**Reply by joshuahansel** [reply]
Yes, copying the module would cause problems. So just to be clear, you no longer get the not-registered error and your immediate issues have been resolved?

**Reply by yifanm4** [reply]
Now I am trying to create a auxkernel for the time derivative term for my auxvariable called conv. 

At first I tried to write a code similar to the example 3, but the code doesn't run because:
computeQpResidual() and computeQpJacobian() never been override even I already declared them in my C file. 

After that, I switched to add ADTimeDerivativeAux into my AuxKernel, but it does make it clear that aux variable isn't currently supported in Auxiliary System (only in non-linear). **So is my auxvaribale consider non-linear if I just define it as:**
[AuxVariables]
  [conv]
    order = CONSTANT
    family = MONOMIAL
  []
[]
**And also, I am not sure what does functor do. it gives the following error if I don't skip it.** 
*** ERROR ***
The parameter "functor" is being retrieved before being set.

**Do I just set it to vel_z, as the value I tried to couple with?**

**Reply by yifanm4** [reply]
> Yes, copying the module would cause problems. So just to be clear, you no longer get the not-registered error and your immediate issues have been resolved?

Yes, the non-registered error has gone. But it does bring more problems as I just listed.

**Reply by joshuahansel** [reply]
`AuxVariables` are for quantities that can be directly computed from other quantities, not for the variables that you want to solve for (those should be `Variables`). For example, `ADTimeDerivativeAux` sets the value of the provided aux variable to be equal to a time derivative; it doesn't add a time derivative term to the nonlinear system.

**Reply by yifanm4** [reply]
> `AuxVariables` are for quantities that can be directly computed from other quantities, not for the variables that you want to solve for (those should be `Variables`). For example, `ADTimeDerivativeAux` sets the value of the provided aux variable to be equal to a time derivative; it doesn't add a time derivative term to the nonlinear system.

So, do you mean that I need to declare a **dust_conv** as a Variable? But the Kernel itself won't allow me to use the calculated **velocity** for coupling.

But if I use an Auxkernel to get the calculated **velocity,** and AuxKernel need to work with AuxVariable, which leaves me no choice but to declare **dust_conv** as AuxVariable. 

How can I solve this paradox?

**Reply by joshuahansel** [reply]
> So, do you mean that I need to declare a dust_conv as a Variable?

Yes, if you're adding an equation to your nonlinear system (your scalar advection equation), you need a variable, not an aux variable.

> But the Kernel itself won't allow me to use the calculated velocity for coupling.

What kernel are you talking about? Did you try the one I attached earlier? That one should couple in the velocity variable from the `FlowChannel1Phase` component.

**Reply by yifanm4** [reply]
> > So, do you mean that I need to declare a dust_conv as a Variable?
> 
> Yes, if you're adding an equation to your nonlinear system (your scalar advection equation), you need a variable, not an aux variable.
> 
> > But the Kernel itself won't allow me to use the calculated velocity for coupling.
> 
> What kernel are you talking about? Did you try the one I attached earlier? That one should couple in the velocity variable from the `FlowChannel1Phase` component.

I tried the ones you attached earlier. Now It does help me couple the velocity variable from the FlowChannel1Phase component and the solutions do converge. Thank you so much

**Reply by yifanm4** [reply]
Hello,

Here is a new question: 

**How can I add one more variable to be advected within the TH module?** 

For Step 1 and 2 from the TH tutorial, I don't need to use junctions, but rather define my own boundary conditions, and it works as expected. But if I need to add junctions to connect different components together, it doesn't seem to pass the information from the previous component to the next component.

**Reply by joshuahansel** [reply]
> How can I add one more variable to be advected within the TH module?

This question seems to be different from your explanation. I'll answer what I think you're asking with your post. I'm guessing that you are asking how to have your new scalar transport equation carry across a junction.

I expect this to be a decent amount of effort. This is going to depend on your spatial discretization for your new equation. For example, what type of finite elements are you using for your new variable?

**Reply by yifanm4** [reply]
> > How can I add one more variable to be advected within the TH module?
> 
> This question seems to be different from your explanation. I'll answer what I think you're asking with your post. I'm guessing that you are asking how to have your new scalar transport equation carry across a junction.
> 
> I expect this to be a decent amount of effort. This is going to depend on your spatial discretization for your new equation. For example, what type of finite elements are you using for your new variable?

I am not sure about what type of finite elements I need to use. Is there anyway to determine it? I assume it will be 1D non-linear. 

Sorry for the confusion about the question. Let me clarify it. I am actually thinking about three different ways of solving this problem. 

1. As you said, define a new variable from the input file and have my new solar transport equation carry it through the junction.  
2. I am also trying to study the source code to create a new variable in the module from the source code similar to the behavior of rhoEA, since it's similar to what I wanted to achieve. Here is what I copied from FlowModelSinglePhase: Total energy equation =(transient term + remaining terms[advection, wall heating, work from body force, etc])
3. Maybe I can create my own mesh which is almost identical to the original one (with the joints removed) and only map the calculated velocity into this new mesh, and then I can easily define the kernel and boundary conditions (since all the junction will be removed in this new mesh) But I am not sure how it can be done, should I try Multi-APP? 

Please let me know what you think about these three approaches

**Reply by joshuahansel** [reply]
I accidentally deleted my original response... I just wanted to move it to be in this thread.

The type of finite elements is determined by the `family` and `order` parameters in your `[Variables]` block. Default is linear Lagrange. That's a "nodal" variable (Its DoFs live at nodes).

Anyway, to help you, I need to know more details. What junction are you trying to use? Are the channels in line at the junction and one-to-one, or do the junctions have many connections? Are the channels parallel? Are they curved?

**Reply by yifanm4** [reply]
> Anyway, to help you, I need to know more details. What junction are you trying to use? Are the channels in line at the junction and one-to-one, or do the junctions have many connections? Are the channels parallel? Are they curved?

I think the linear Lagrange is good for now. 

Since it's just the new defined variable pass through the same loop as TH module Tutorial defined, won't the junctions the same as what has been used from the input files? One type to take care of the parallel connections (0 degree) and the other one for the horizontal and vertical (90 degrees) connections. And it should be one to one connection between channels. 

From the input file **06_custom_closure.i**, it uses the following two junction types:
1. JunctionParallelChannels1Phase
2. VolumeJunction1Phase

**Reply by joshuahansel** [reply]
Ok, so these are both "volume" junctions. If you only had one-to-one junctions, we'd probably just make a constraint between ends of the junction that says that your new variable is equal. But with a volume junction, you'll need to add a new scalar variable that tracks the amount of your new variable in the 0D junction volume. Then you'll need to add scalar kernels corresponding to each connection to the junction and then corresponding boundary conditions on the channels.

**Reply by yifanm4** [reply]
Ok, sounds good. I understand why it matters to know the junction types now. For my case, I just need the information to pass by the junction without changing (new variable is equal to the old variable) . Which junction type I should use? And how do I make a constraint between ends of the junction to make it work?

**Reply by joshuahansel** [reply]
Ok, if you want the quantity to pass by the junction without noticing, it sounds like you need to make a constraint to make your new variable equal on the connected boundaries. I don't which one would be best for your situation, but it looks like there are a few aimed at constraining equal values between different boundaries, so you should browse those.

As for junction type, it doesn't matter - your implementation for this won't be in a junction anyway. Choose whichever are best for your other flow equations. I was mostly asking to see if you have branching flow or not, but you do not, though you have bends.

**Reply by yifanm4** [reply]
> Ok, if you want the quantity to pass by the junction without noticing, it sounds like you need to make a constraint to make your new variable equal on the connected boundaries. I don't which one would be best for your situation, but it looks like there are a few aimed at constraining equal values between different boundaries, so you should browse those.
> 
> As for junction type, it doesn't matter - your implementation for this won't be in a junction anyway. Choose whichever are best for your other flow equations. I was mostly asking to see if you have branching flow or not, but you do not, though you have bends.

Alright, thank you for the explanation. There are two more things I try to figure out.
1. For the **ConservativeAdvectionScalar** Kernel you helped me created. If there is 1-D straight line, I can just set a fixed orientation, not an issue. But since I do have bends in the loop and the orientation needs to be changed with the components set up. **Is there a way allowing me to change the orientation dynamically?** 

2. **How to set up the boundary conditions for a loop like this?**

**Reply by joshuahansel** [reply]
There is `Elbow1Phase` for making a circular curved channel, which you can connect to straight channels. However, see the recent discussion https://github.com/idaholab/moose/discussions/22787: the input can be very confusing. There isn't any arbitrary curved channel ability yet.

I'm not sure what you mean by boundary conditions. The flow boundary conditions are used in exactly the same way as for a straight channel.

**Reply by yifanm4** [reply]
> There is `Elbow1Phase` for making a circular curved channel, which you can connect to straight channels. However, see the recent discussion #22787: the input can be very confusing. There isn't any arbitrary curved channel ability yet.
> 
> I'm not sure what you mean by boundary conditions. The flow boundary conditions are used in exactly the same way as for a straight channel.

I tried to use junctions to connect two components in my sub_app, which has the app_type = ThermalHydraulicsApp, but the following error message pops up:

**add_component is not registered to build BoundaryCondition derived objects**

Any suggestions on how to fix this?

**Reply by yifanm4** [reply]
Also, is there a way for me to extract my generated mesh from the input file out as a .e file and reuse for some other applications later?

**Reply by joshuahansel** [reply]
> add_component is not registered to build BoundaryCondition derived objects

I'm guessing you've put a `BoundaryCondition` object in your `[Components]` block. It needs to be in a `[BoundaryConditions]` block.

**Reply by joshuahansel** [reply]
> Also, is there a way for me to extract my generated mesh from the input file out as a .e file and reuse for some other applications later?

Do you mean a mesh generated from components or a mesh generated from the mesh generator system? If from components, I'm not as sure, but I think you can just do a run with zero time steps and then use the exodus output file you get to load a mesh. Maybe you can restrict the output to exclude all of the non-mesh stuff. If using the mesh generator system, you can make an input file with just the mesh generators and then run your application with the --mesh-only flag to just output the mesh to an exodus file.

**Reply by yifanm4** [reply]
> > Also, is there a way for me to extract my generated mesh from the input file out as a .e file and reuse for some other applications later?
> 
> Do you mean a mesh generated from components or a mesh generated from the mesh generator system? If from components, I'm not as sure, but I think you can just do a run with zero time steps and then use the exodus output file you get to load a mesh. Maybe you can restrict the output to exclude all of the non-mesh stuff. If using the mesh generator system, you can make an input file with just the mesh generators and then run your application with the --mesh-only flag to just output the mesh to an exodus file.

Thank you so much! I do get it working now. Here is one more question related to the loop. 

I try to set up a simple square loop without setting any boundary conditions. I try to set a ConstantIC on the left side and initialize the dust convection with a constant value. However, since there is no boundaries to stop the dust from going the opposite direction, it does flow in the non-physical direction. (Since the dust is carried by the fluid, thus it should flow in the same direction of the fluid). Any suggestions on how can I restrict it to flow in the correct direction?

**Reply by joshuahansel** [reply]
So you have your main fluid going in one direction around the loop, but your dust is going the opposite direction? Is it opposite in all of the four sides or just some? I don't know what this would have to do with the lack of boundaries. It sounds more like some sign issue in your equations or some problem at your junctions.

**Reply by yifanm4** [reply]
> So you have your main fluid going in one direction around the loop, but your dust is going the opposite direction? Is it opposite in all of the four sides or just some? I don't know what this would have to do with the lack of boundaries. It sounds more like some sign issue in your equations or some problem at your junctions.

Sorry for the confusion, what I mean is that the dust is actually going in both directions, behave somewhat like a numerical diffusion. I am trying to get rid of the movement of the dust in the opposite direction. Shall I show you the input files?

**Reply by joshuahansel** [reply]
I don't know that the input file would tell me much. This is probably something where we'd need to look at your implementation (equations, spatial discretization, etc.). I'd like to avoid that if possible.

I recommend you start with a simpler problem if possible: a single channel with an inlet and an outlet. Make sure to add boundary conditions for your dust equation, according to your spatial discretization (supposing there is some). Put an initial dust concentration in the middle of the channel and verify it goes only in the correct direction. If it doesn't, you need to troubleshoot this: make sure you can prove on paper that it should only go in the correct direction. If it does work, then replace your outlet with a junction connected to another channel and then an outlet. If this fails, then there's something wrong in your junction formulation.
